src/Assistant.ts

```diff
@@ -0,0 +1,387 @@
+
+    return preparedArgs;
+  }
+}
+
+/**
+ * `isAssistantEvent()` determines if incoming event is a supported
+ * Assistant event type.
+ */
+export function isAssistantEvent(args: AnyMiddlewareArgs): boolean {
+  return ASSISTANT_PAYLOAD_TYPES.has(args.payload.type);
+}
+
+/**
+ * `matchesConstraints()` determines if the incoming event payload
+ * is related to the Assistant.
+ */
+export function matchesConstraints(args: AnyMiddlewareArgs): args is AllAssistantMiddlewareArgs {
+  return args.payload.type === 'message' ? isAssistantMessage(args.payload) : true;
+}
+
+/**
+ * `isAssistantMessage()` evaluates if the message payload is associated
+ * with the Assistant container.
+ */
+export function isAssistantMessage(payload: AnyMiddlewareArgs['payload']): boolean {
+  const isThreadMessage = 'channel' in payload && 'thread_ts' in payload;
+  const inAssistantContainer = ('channel_type' in payload && payload.channel_type === 'im') &&
+  (!('subtype' in payload) || payload.subtype === 'file_share');
+  return isThreadMessage && inAssistantContainer;
+}
+
+/**
+ * `validate()` determines if the provided AssistantConfig is a valid configuration.
+ */
+export function validate(config: AssistantConfig): void {
+  // Ensure assistant config object is passed in
+  if (typeof config !== 'object') {
+    const errorMsg = 'Assistant expects a configuration object as the argument';
+    throw new AssistantInitializationError(errorMsg);
+  }
+
+  // Check for missing required keys
+  const requiredKeys: (keyof AssistantConfig)[] = ['threadStarted', 'userMessage'];
+  const missingKeys: (keyof AssistantConfig)[] = [];
+  requiredKeys.forEach((key) => { if (config[key] === undefined) missingKeys.push(key); });
+
+  if (missingKeys.length > 0) {
+    const errorMsg = `Assistant is missing required keys: ${missingKeys.join(', ')}`;
+    throw new AssistantInitializationError(errorMsg);
+  }
+
+  // Ensure a callback or an array of callbacks is present
+  const requiredFns: (keyof AssistantConfig)[] = ['threadStarted', 'userMessage'];
+  requiredFns.forEach((fn) => {
+    if (typeof config[fn] !== 'function' && !Array.isArray(config[fn])) {
+      const errorMsg = `Assistant ${fn} property must be a function or an array of functions`;
+      throw new AssistantInitializationError(errorMsg);
+    }
+  });
+}
+
+/**
+ * `processAssistantMiddleware()` invokes each callback for the given event
+ */
+export async function processAssistantMiddleware(
+  args: AllAssistantMiddlewareArgs,
+  middleware: AssistantMiddleware,
+): Promise<void> {
+  const { context, client, logger } = args;
+  const callbacks = [...middleware] as Middleware<AnyMiddlewareArgs>[];
+  const lastCallback = callbacks.pop();
+
+  if (lastCallback !== undefined) {
+    await processMiddleware(
+      callbacks, args, context, client, logger,
+      async () => lastCallback({ ...args, context, client, logger }),
+    );
+  }
+}
+
+function selectToken(context: Context): string | undefined {
```

Comment:
@seratch I realized as I was writing this that `selectToken` is only used for `say` in `App.ts`. Does that mean that whatever token comes along with the `client` is the one that we want to use? If so, I think I can remove this and the additional token passing that occurs in the utility functions.

Reply:
when enabling the agents & assistants feature, the bolt-js app should have the bot scope to write messages for sure and the primary token in the context / client should be a valid bot token, not a human user token. Thus, for this case, I don't see the necessity to have selectToken within this middleware.

Reply:
Will remove it then